In [30]:
from sklearn.feature_extraction.text import CountVectorizer
from konlpy.tag import Okt
import pandas as pd

## tokenized된 리뷰 데이터 전처리 보완하기
* 불러오기
* null 값 있는 컬럼 제거하기
* 유의어 일원화 보완
* 갱신된 불용어 사전 적용

In [63]:
df = pd.read_csv("../../data/naver_shopping_tokenized_review.csv", encoding ='utf-8', index_col = 0)
df[100:120]

Score                                             Review  y  \
100      1                                      생각외로 모기가 안잡혀요  0   
101      2  고구마가 외이케 말랏나요 저나하려다 바빠깜팍 햇는데 서너번 주문한적잇는데 이런적업엇...  0   
102      2  (바코드가 왜 손상되서 오는건지 알려주세요) 10년 넘게 써오던 제품이라 항상 만족...  0   
103      5                       대용량 박스 사이즈 잘 체크하고 사세요 많다 큽니다  1   
104      4                                         시원하구 좋아용!!  1   
105      5   제품주 좋지만 특히 배송날짜 칼이네요 이사간날 정확하게 배달 설치해주시구 감사합니다~~  1   
106      5                       늘 쓰던것~~여름이라 그런지 지속력이 좀떨어지네요ㅜ  1   
107      5                                    가격이 저렴하다합니다 ㅎㅎㅎ  1   
108      1  상품가치 없는크기를 보내셨어요 한입사이즈크기에 3/1은 상했구요 이건좀 아니다싶네요...  0   
109      2                                    쓸데가 하나없음 효과일도없음  0   
110      4  저렴하다하게 받아서 기분은좋아요 근데 가방에 실리콘 같은게 좀 묻어있더라구요.. 그...  1   
111      5                          빠른배송감사합니다. 역시 헤라가 짱입니다~^^  1   
112      5  받자마자 아이랑 같이 아이스아이보리 만들어먹었는데 실리콘으로 되어서 빼기도 편하고 ...  1   
113      1                               어이가없네요 셈플한개가 비어서 왔네요  0   
114      2                                  작업하기까다로움 효과도 체감안됨  0   
115      4                              만족합니다. 날씬하다핏 아닌 레귤러네요  1   
116      5               왜 승무원 스프레이라고하는지 알겠네요 짱이에요. 흡수력 광택 @@  1   
117      5  연어초밥, 연어장 만들어 먹었는데 너무 맛있어요! 내일은 연어 스테이크랑 까나페 해...  1   
118      5                   상판이 조금 약하긴한데 일부러 누르지 않는한 크고 좋아요.  1   
119      5                                   시원해서 잘쓸거같아요 여름에!  1   

                                           tokenized_1  \
100                                       생각 외로 모기 잡혀요   
101  고구마 외이케 말랏나 나하 바빠깜팍 는데 한적 잇는데 적업 는데 사이즈 정도 아니라...   
102                 바코드 손상 되서 알려주세요 넘게 써오던 제품 항상 만족합니다   
103                            용량 박스 사이즈 체크 사세요 많다 큽니다   
104                                           시원하구 좋아용   
105              제품 좋지만 배송 날짜 이사 정확하게 배달 설치 해주시구 감사합니다   
106                                쓰던것 여름 그런지 지속 떨어지네요   
107                                        가격 저렴하다 합니다   
108    상품 가치 없는 크기 보내셨어요 입사 크기 했구요 이건 아니다싶네요 정가 할인 크던데   
109                                    쓸데가 없음 효과 일도 없음   
110        저렴하다하게 받아서 기분 좋아요 가방 실리콘 같은게 묻어있더라구요 괜찬 같아요   
111                              빠른 배송 감사합니다 역시 헤라 입니다   
112         받자마자 아이스 아이보리 만들어 먹었는데 실리콘 되어서 빼기 편하고 좋았어요   
113                                         없네요 비어 왔네요   
114                                 작업 하기 까다로 효과 체감 안됨   
115                                 만족합니다 날씬하다 아닌 레귤러네   
116                            승무원 스프레이 하는지 알겠네요 흡수 광택   
117      연어초밥 연어 만들어 먹었는데 맛있어요 내일 연어 스테이크 까나페 먹으려고 합니다   
118                        판이 약하긴한데 일부러 누르지 않는한 크고 좋아요   
119                                    시원해서 잘쓸거 같아요 여름   

                                           tokenized_2  
100                                       생각 외로 모기 잡혀요  
101  고구마 외이케 말랏나 나하 려 바빠깜팍 햇 는데 서너 번 문 한적 잇는데 적업 엇 ...  
102  ( 바코드 손상 되서 오는 건 알려주세요 ) 10년 넘게 써오던 제품 이라 항상 만...  
103                       용량 박스 사이즈 잘 체크 하고 사세요 많다 큽니다  
104                                        시원하구 좋아용 !!  
105  제품 좋지만 특히 배송 날짜 칼 이네 이사 간 날 정확하게 배달 설치 해주시구 감사...  
106                    늘 쓰던것 ~~ 여름 이라 그런지 지속 력 떨어지네요 ㅜ  
107                                    가격 저렴하다 합니다 ㅎㅎㅎ  
108  상품 가치 없는 크기 보내셨어요 입사 크기 3/1 은 했구요 이건 아니다싶네요 정가...  
109                                    쓸데가 없음 효과 일도 없음  
110  저렴하다하게 받아서 기분 은 좋아요 근데 가방 실리콘 같은게 묻어있더라구요 .. 말...  
111                      빠른 배송 감사합니다 . 역시 헤라 짱 입니다 ~^^  
112       받자마자 랑 아이스 아이보리 만들어 먹었는데 실리콘 되어서 빼기 편하고 좋았어요  
113                                   없네요 셈 플 비어 서 왔네요  
114                               작업 하기 까다로 움 효과 체감 안됨  
115                             만족합니다 . 날씬하다 핏 아닌 레귤러네  
116            승무원 스프레이 라고 하는지 알겠네요 짱 이에요 . 흡수 력 광택 @@  
117  연어초밥 , 연어 장 만들어 먹었는데 너무 맛있어요 ! 내일 은 연어 스테이크 랑 ...  
118                      판이 약하긴한데 일부러 누르지 않는한 크고 좋아요 .  
119                                  시원해서 잘쓸거 같아요 여름 !

In [64]:
df.isnull().sum()

Score            0
Review           0
y                0
tokenized_1    428
tokenized_2     15
dtype: int64

In [65]:
df = df[df['tokenized_1'].notnull() & df['tokenized_2'].notnull()]
df.isnull().sum()

Score          0
Review         0
y              0
tokenized_1    0
tokenized_2    0
dtype: int64

In [87]:
wordList = ['만족합니다','만족스럽습니다','만족스럽네요','만족합니다','만족스러워요']
change = '만족해요'

for word in wordList:
    df['tokenized_1'] = df['tokenized_1'].str.replace(word, change, regex = True)
    df['tokenized_2'] = df['tokenized_2'].str.replace(word, change, regex = True)

In [119]:
wordList = ['좋구요','좋네요','좋습니다','좋아해요','좋아합니다','좋아요요']
change = '좋아요'

for word in wordList:
    df['tokenized_1'] = df['tokenized_1'].str.replace(word, change, regex = True)
    df['tokenized_2'] = df['tokenized_2'].str.replace(word, change, regex = True)

In [118]:
wordList = ['안좋아요요']
change = '안좋아요'

for word in wordList:
    df['tokenized_1'] = df['tokenized_1'].str.replace(word, change, regex = True)
    df['tokenized_2'] = df['tokenized_2'].str.replace(word, change, regex = True)

In [102]:
wordList = ['에드', '맘에듭니']
change = '마음에 듭니다'

for word in wordList:
    df['tokenized_1'] = df['tokenized_1'].str.replace(word, change, regex = True)
    df['tokenized_2'] = df['tokenized_2'].str.replace(word, change, regex = True)

In [160]:
wordList = ['좋아여', '좋아용','좋아하시네요','좋아요하시네요','좋아요요']
change = '좋아요'

for word in wordList:
    df['tokenized_1'] = df['tokenized_1'].str.replace(word, change, regex = True)
    df['tokenized_2'] = df['tokenized_2'].str.replace(word, change, regex = True)

In [115]:
wordList = ['맛있었어요']
change = '맛있어요'

for word in wordList:
    df['tokenized_1'] = df['tokenized_1'].str.replace(word, change, regex = True)
    df['tokenized_2'] = df['tokenized_2'].str.replace(word, change, regex = True)

In [152]:
wordList = ['예쁘다','이쁘고','예쁘고']
change = '예뻐요'

for word in wordList:
    df['tokenized_1'] = df['tokenized_1'].str.replace(word, change, regex = True)
    df['tokenized_2'] = df['tokenized_2'].str.replace(word, change, regex = True)

In [153]:
wordList = ['좋았어요','좋으네요']
change = '좋아요'

for word in wordList:
    df['tokenized_1'] = df['tokenized_1'].str.replace(word, change, regex = True)
    df['tokenized_2'] = df['tokenized_2'].str.replace(word, change, regex = True)

In [154]:
wordList = ['느리네요','느리고','느려서']
change = '느려요'

for word in wordList:
    df['tokenized_1'] = df['tokenized_1'].str.replace(word, change, regex = True)
    df['tokenized_2'] = df['tokenized_2'].str.replace(word, change, regex = True)

In [69]:
# stopwords = pd.read_csv("https://raw.githubusercontent.com/yoonkt200/FastCampusDataset/master/korean_stopwords.txt").values.tolist()
# temp = []
# for s in stopwords:
#     temp.append(s[0])

f = open("../../data/stopwords_self_improved.txt", 'r', encoding= 'utf-8')

lines = f.readlines()
for idx, l in enumerate(lines):
    lines[idx] = l.strip()

stopwords = []
for l in lines:
    stopwords.append(l)



In [70]:
import time
from tqdm import tqdm

for stopword in tqdm(stopwords):
    df.replace(stopword, '', regex = True, inplace = True)

100%|██████████| 382/382 [04:18<00:00,  1.48it/s]


In [158]:
df.to_csv("../../data/naver_shopping_tokenized_review_improved.csv", sep = ",", encoding = 'utf-8')

## 토큰화 함수 정의
* 원래는 CounterVectorizer나 TFIDF 하이퍼파라미터에 해당 tokenizer_1, tokenizer_2 함수를 tokenizer로 지정하는 방법이 기본적인 방법
* 시간을 효율적으로 쓰기 위해 tokenizer_1, tokenizer_2 함수를 적용한 컬럼 tokenized_1, tokenized_2를 df에 추가함
* 명사, 형용사, 동사 + 1글자 이상만 추출하는 경우 - tokenized_1 리뷰 데이터 활용
* 형태소 모두 사용 - tokenized_2 리뷰 데이터 활용
* vectorized 클래스에 들어갈 tokenizer는 띄어쓰기 기준으로 split하는 기능만 적용

In [72]:
def tokenizer(text):
    return text.split()

### 토큰화된 데이터 프레임 불러오기

In [161]:
df[250:270]

Score                                             Review  y  \
250      5  지방이라서 이 늦을줄 알았는데 도 빠르고 제품도 생각한것처럼 예쁘고 따갑다끔 특히나...  1   
251      1       상품은 늘 쓰던거라 했는데요 제조일자 6개월이상 된거 보내주셨네요 기분 나쁩니다  0   
252      1  고들빼기라 되있는데 깻잎이?? 제가 깻잎을 좋아하고 가격도 비슷하고 귀차나서 다시 ...  0   
253      2                                         붓기 넘넘 잘빧아요  0   
254      2                        물건을 얼려놓으면 앞 다쏠리고 힘아리가 도 없네요  0   
255      2          신맛이 나요.. 무친지 좀 오래된듯.. 산거 후회합니다. 다 버려야 할듯   0   
256      2  물빠짐 물빠짐 심합니다 실보프라기도 심해서 한참 제게했고 중간중간 올플림도 있어요 ...  0   
257      5             소리도 잘나오고 가성비 정말 좋아요 드라이브 기능도 있어서 너무 만족  1   
258      1  정전기 대박.... 기장가 긴건 내 키가 작아서라고쳐도 정전기는 왜이렇게 심한건지 ...  0   
259      5                                     맛있어요 고급스러워보입니다  1   
260      5  좋아요 아이 연습용 했습니다 저는 바이올린은 아예 몰라서 유튜브영상 송진바르는법 이...  1   
261      5                                         열전달도잘괴고좋아요  1   
262      1                           너무 크네요 생각보다 크게 나오니 주의하셔요  0   
263      1                남편 선물로 사줬는데... 하루만에 벗겨졌어요.... 비추입니다  0   
264      2                               필터 마크가 두칸밖에안뜨요.정품맞아요  0   
265      5                                     괜찬습니다^^ 잘쓰고있어요  1   
266      4                                           빨랐어요 좋아요  1   
267      2               상품은 이쁜데 갓이 다 찌그러져서 왔어요 포장을 좀더 신경써주세요  0   
268      2                                         이 넘 느리네요..  0   
269      1                            앙고라 털이 엄청빠져요 옷에 다 붙어나와요  0   

                                           tokenized_1  \
250      지방  늦을줄 알았는데  빠르고 제품 생각 예뻐요 따갑다  기사 친절하셔서 좋아요   
251           상품 쓰던거라  했는데요 제조 일자 개월 된거 보내주셨네요 기분 나쁩니다   
252  고들빼기 되있는데 깻잎 깻잎 좋아요하고 가격 비슷하고 나서 다시 환불 교환 요청 합...   
253                                             붓기 잘빧아   
254                                물건 얼려놓으면 쏠리고 아리 없네요   
255                       신맛 무친지 오래된듯 산거 후회 합니다 버려야 할듯   
256   빠짐 빠짐 심합니다 실보 프라 기도 심해서 한참 했고 중간 중간 있어요 가격 별로 다예   
257                   소리 잘나오고 가성 정말 좋아요 드라이브 기능 있어서 만족   
258  정전기 대박 기장 긴건 작아 서라고 쳐도 정전기 심한 건지 만지기만 하면 따닥따다닥...   
259                                    맛있어요 고급스러워 보입니다   
260  좋아요 연습  했습니다 바이올린 아예 몰라서 유튜브 영상 송진 바르는 튜닝법 검색 ...   
261                                         열전달 괴고 좋아요   
262                               크네요 생각 크게 나오니 주의 하셔요   
263                         남편 선물 사줬는데 하루 벗겨졌어요 비추 입니다   
264                                       필터 마크 요정 맞아요   
265                                        찬습니다 잘쓰고있어요   
266                                           빨랐어요 좋아요   
267                     상품 이쁜데 찌그러져서 왔어요 포장 좀더 신경 써주세요   
268                                                느려요   
269                                     앙고라 빠져요 붙어 나와요   

                                           tokenized_2  
250  지방 이라서  늦을줄 알았는데  빠르고 제품 생각 처럼 예뻐요 따갑다 끔 특히  기...  
251     상품 은 늘 쓰던거라  했는데요 제조 일자 6 개월 된거 보내주셨네요 기분 나쁩니다  
252  고들빼기 되있는데 깻잎 ?? 깻잎 좋아요하고 가격 비슷하고 귀 나서 다시 환불 이나...  
253                                         붓기 넘 넘 잘빧아  
254                                물건 얼려놓으면 쏠리고 아리 없네요  
255            신맛 나요 .. 무친지 오래된듯 .. 산거 후회 합니다 . 버려야 할듯  
256  빠짐 빠짐 심합니다 실보 프라 기도 심해서 한참 했고 중간 중간 올 플 림 있어요 ...  
257                소리 잘나오고 가성 정말 좋아요 드라이브 기능 있어서 너무 만족  
258  정전기 대박 .... 기장 긴건 내 키 작아 서라고 쳐도 정전기 는 이렇게 심한 건...  
259                                    맛있어요 고급스러워 보입니다  
260  좋아요 연습 용  했습니다 는 바이올린 은 아예 몰라서 유튜브 영상 송진 바르는 법...  
261                                       열전달 잘 괴고 좋아요  
262                         너무 크네요 생각 보다 크게 나오니 주의 하셔요  
263             남편 선물 사줬는데 ... 하루 만에 벗겨졌어요 .... 비추 입니다  
264                             필터 마크 칸 밖에 뜨요 . 정품 맞아요  
265                                   괜 찬습니다 ^^ 잘쓰고있어요  
266                                           빨랐어요 좋아요  
267                 상품 은 이쁜데 갓 찌그러져서 왔어요 포장 좀더 신경 써주세요  
268                                           넘 느려요 ..  
269                                앙고라 털 엄청 빠져요 붙어 나와요

In [162]:
print(df.isnull().sum())
df.info()

Score          0
Review         0
y              0
tokenized_1    0
tokenized_2    0
dtype: int64
<class 'pandas.core.frame.DataFrame'>
Int64Index: 199572 entries, 0 to 199999
Data columns (total 5 columns):
 #   Column       Non-Null Count   Dtype 
---  ------       --------------   ----- 
 0   Score        199572 non-null  int64 
 1   Review       199572 non-null  object
 2   y            199572 non-null  int64 
 3   tokenized_1  199572 non-null  object
 4   tokenized_2  199572 non-null  object
dtypes: int64(2), object(3)
memory usage: 9.1+ MB


## DTM + GridSearchCV + tokenizer_1
* countervectorizer
* GridSearchCV
* Tokenizer1

In [209]:
from sklearn.model_selection import train_test_split
y = df['y']
X_train_texts, X_test_texts, y_train, y_test = train_test_split(df['tokenized_1'], y, test_size = 0.2, random_state = 0)

In [210]:
vect = CountVectorizer(min_df = 3, ngram_range=(1,2), tokenizer = tokenizer)
X_train_tf = vect.fit_transform(X_train_texts)
X_test_tf = vect.transform(X_test_texts)

c:\Users\taeri\miniconda3\envs\ds_study\lib\site-packages\sklearn\feature_extraction\text.py:528: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(


In [211]:
vocablist_DTM_tk1 = [word for word, number in sorted(vect.vocabulary_.items(), key = lambda x:x[1])]

In [215]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.model_selection import GridSearchCV

lr = LogisticRegression(random_state = 0, solver = 'liblinear', max_iter = 1000)
params = { 'C': [0.6, 0.8, 1 ,3.5, 4.5 ] }
grid_cv = GridSearchCV(lr , param_grid=params , cv=3 ,scoring='accuracy', verbose=1 )
grid_cv.fit(X_train_tf , y_train)
print(grid_cv.best_params_ , round(grid_cv.best_score_,4))

Fitting 3 folds for each of 5 candidates, totalling 15 fits
{'C': 0.8} 0.8947


In [216]:
best_estimator_DTM_tk1 = grid_cv.best_estimator_
y_pred = best_estimator_DTM_tk1.predict(X_test_tf)

In [217]:
print("accuracy: %.2f" % accuracy_score(y_test, y_pred))
print("Precision : %.3f" % precision_score(y_test, y_pred))
print("Recall : %.3f" % recall_score(y_test, y_pred))
print("F1 : %.3f" % f1_score(y_test, y_pred))

accuracy: 0.90
Precision : 0.905
Recall : 0.891
F1 : 0.898


In [218]:
import pickle

with open('DTM_tokenizer1_gridsearch_done_improved.pickle','wb') as fw:
    pickle.dump(best_estimator_DTM_tk1, fw)

In [219]:
coefficients = best_estimator_DTM_tk1.coef_.tolist()

sorted_coefficients = sorted(enumerate(coefficients[0]), key=lambda x:x[1], reverse=True)
# coefficients(계수)가 큰 값부터 내림차순으로 정렬

print('긍정적인 단어 Top 20 (높은 평점과 상관관계가 강한 단어들)')
for word_num, coef in sorted_coefficients[:20]:
  print('{0:}({1:.3f})'.format(vocablist_DTM_tk1[word_num], coef))

긍정적인 단어 Top 20 (높은 평점과 상관관계가 강한 단어들)
맛있어요(2.965)
튼튼합니다(2.831)
만족해요(2.626)
잘쓰고있어요(2.510)
좋아요(2.458)
최고(2.399)
튼튼하네요(2.390)
존맛(2.305)
예쁘네요(2.300)
편(2.288)
예뻐요(2.262)
좋아요하시네요(2.229)
괜찮네요(2.225)
빠르고(2.215)
귀엽다(2.214)
잘썼어요(2.171)
딱이에요(2.156)
편합니다(2.134)
튼튼하고(2.107)
고급스러워요(2.103)


In [220]:
print('\n부정적인 단어 Top 20 (낮은 평점과 상관관계가 강한 단어들)')
for word_num, coef in sorted_coefficients[-20:][::-1]: 
  print('{0:}({1:.3f})'.format(vocablist_DTM_tk1[word_num], coef))


부정적인 단어 Top 20 (낮은 평점과 상관관계가 강한 단어들)
최악(-3.748)
실망(-3.392)
비추(-3.298)
다시는(-2.999)
약(-2.759)
좋아요 좋아요(-2.741)
맛있어요 맛있어요(-2.708)
맛없어요(-2.683)
심하네요(-2.660)
그래요(-2.545)
환불(-2.522)
별로(-2.454)
안됩니다(-2.411)
느려요(-2.383)
떨어집니다(-2.328)
불친절(-2.300)
안좋아요(-2.292)
불편합니다(-2.280)
반품(-2.279)
안됨(-2.259)


## TFIDF + GridSearchCV + tokenizer_1
* TFIDF
* GridSearchCV
* Tokenizer1

In [221]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_vectorizer = TfidfVectorizer(ngram_range=(1,2), min_df = 3, max_df = 0.9, tokenizer = tokenizer)
tfidf_vectorizer.fit(X_train_texts)
tfidf_matrix_train = tfidf_vectorizer.transform(X_train_texts)
tfidf_matrix_test = tfidf_vectorizer.transform(X_test_texts)

c:\Users\taeri\miniconda3\envs\ds_study\lib\site-packages\sklearn\feature_extraction\text.py:528: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(


In [222]:
vocablist_TFIDF_tk1 = [word for word, number in sorted(tfidf_vectorizer.vocabulary_.items(), key = lambda x:x[1])]

In [223]:
lr = LogisticRegression(random_state = 0, solver = 'liblinear', max_iter = 1000)
params = { 'C': [0.6, 0.8, 1 ,3.5, 4.5 ] }
grid_cv = GridSearchCV(lr , param_grid=params , cv=3 ,scoring='accuracy', verbose=1 )
grid_cv.fit(tfidf_matrix_train , y_train)
print(grid_cv.best_params_ , round(grid_cv.best_score_,4))

Fitting 3 folds for each of 5 candidates, totalling 15 fits
{'C': 3.5} 0.8967


In [224]:
best_estimator_TFIDF_tk1 = grid_cv.best_estimator_
y_pred = best_estimator_TFIDF_tk1.predict(tfidf_matrix_test)

In [225]:
print("accuracy: %.2f" % accuracy_score(y_test, y_pred))
print("Precision : %.3f" % precision_score(y_test, y_pred))
print("Recall : %.3f" % recall_score(y_test, y_pred))
print("F1 : %.3f" % f1_score(y_test, y_pred))

accuracy: 0.90
Precision : 0.905
Recall : 0.893
F1 : 0.899


In [226]:
with open('TFIDF_tokenizer1_gridsearch_done_improved.pickle','wb') as fw:
    pickle.dump(best_estimator_TFIDF_tk1, fw)

In [227]:
coefficients = best_estimator_TFIDF_tk1.coef_.tolist()

sorted_coefficients = sorted(enumerate(coefficients[0]), key=lambda x:x[1], reverse=True)
# coefficients(계수)가 큰 값부터 내림차순으로 정렬

print('긍정적인 단어 Top 20 (높은 평점과 상관관계가 강한 단어들)')
for word_num, coef in sorted_coefficients[:20]:
  print('{0:}({1:.3f})'.format(vocablist_TFIDF_tk1[word_num], coef))

긍정적인 단어 Top 20 (높은 평점과 상관관계가 강한 단어들)
좋아요(10.449)
맛있어요(10.234)
만족해요(8.684)
예뻐요(8.546)
최고(7.722)
빠르고(7.312)
튼튼하고(6.832)
만족(6.322)
편(6.097)
걱정 했는데(5.774)
편하고(5.654)
괜찮네요(5.567)
귀엽다(5.538)
강추(5.441)
튼튼합니다(5.404)
항상(5.383)
좋았습니다(5.254)
있어서 좋아요(5.214)
빨랐고(5.109)
듭니(5.085)


In [228]:
print('\n부정적인 단어 Top 20 (낮은 평점과 상관관계가 강한 단어들)')
for word_num, coef in sorted_coefficients[-20:][::-1]: 
  print('{0:}({1:.3f})'.format(vocablist_TFIDF_tk1[word_num], coef))


부정적인 단어 Top 20 (낮은 평점과 상관관계가 강한 단어들)
실망(-13.225)
반품(-11.114)
별로(-11.026)
최악(-10.417)
비추(-9.335)
다시는(-8.819)
그렇다(-7.887)
환불(-7.631)
약(-6.835)
엉망(-6.493)
불편(-6.381)
안좋아요(-6.311)
안됩니다(-6.284)
심하네요(-6.272)
버렸어요(-6.209)
맛없어요(-5.979)
불편합니다(-5.869)
떨어지네요(-5.852)
불량(-5.764)
떨어집니다(-5.758)


## DTM + GridSearchCV + tokenizer_2
* countervectorizer
* GridSearchCV
* Tokenizer2

In [229]:
X_train_texts, X_test_texts, y_train, y_test = train_test_split(df['tokenized_2'], y, test_size = 0.2, random_state = 0)

In [230]:
vect2 = CountVectorizer(min_df = 3, ngram_range=(1,2), tokenizer = tokenizer)
X_train_tf = vect2.fit_transform(X_train_texts)
X_test_tf = vect2.transform(X_test_texts)

c:\Users\taeri\miniconda3\envs\ds_study\lib\site-packages\sklearn\feature_extraction\text.py:528: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(


In [231]:
vocablist_DTM_tk2 = [word for word, number in sorted(vect2.vocabulary_.items(), key = lambda x:x[1])]

In [232]:
lr = LogisticRegression(random_state = 0, solver = 'liblinear', max_iter = 1000)
params = { 'C': [0.6, 0.8, 1 ,3.5, 4.5 ] }
grid_cv = GridSearchCV(lr , param_grid=params , cv=3 ,scoring='accuracy', verbose=1 )
grid_cv.fit(X_train_tf , y_train)
print(grid_cv.best_params_ , round(grid_cv.best_score_,4))

Fitting 3 folds for each of 5 candidates, totalling 15 fits
{'C': 0.6} 0.9111


In [233]:
best_estimator_DTM_tk2 = grid_cv.best_estimator_
y_pred = best_estimator_DTM_tk2.predict(X_test_tf)

print("accuracy: %.2f" % accuracy_score(y_test, y_pred))
print("Precision : %.3f" % precision_score(y_test, y_pred))
print("Recall : %.3f" % recall_score(y_test, y_pred))
print("F1 : %.3f" % f1_score(y_test, y_pred))

accuracy: 0.91
Precision : 0.915
Recall : 0.910
F1 : 0.912


In [234]:
with open('DTM_tokenizer2_gridsearch_done_improved.pickle','wb') as fw:
    pickle.dump(best_estimator_DTM_tk2, fw)

In [235]:
coefficients = best_estimator_DTM_tk2.coef_.tolist()

sorted_coefficients = sorted(enumerate(coefficients[0]), key=lambda x:x[1], reverse=True)
# coefficients(계수)가 큰 값부터 내림차순으로 정렬

print('긍정적인 단어 Top 20 (높은 평점과 상관관계가 강한 단어들)')
for word_num, coef in sorted_coefficients[:20]:
  print('{0:}({1:.3f})'.format(vocablist_DTM_tk2[word_num], coef))

긍정적인 단어 Top 20 (높은 평점과 상관관계가 강한 단어들)
맛있어요(2.674)
만족해요(2.453)
튼튼합니다(2.296)
최고(2.171)
좋아요(2.154)
잘쓰고있어요(2.108)
빠르고(2.075)
잘 왔어요(2.074)
예뻐요(2.021)
괜찮네요(1.991)
예쁘네요(1.923)
만족(1.909)
튼튼하고(1.886)
튼튼하네요(1.881)
편합니다(1.880)
좋았습니다(1.862)
편하다(1.850)
딱이에요(1.841)
좋아요하시네요(1.811)
촉촉(1.807)


In [236]:
print('\n부정적인 단어 Top 20 (낮은 평점과 상관관계가 강한 단어들)')
for word_num, coef in sorted_coefficients[-20:][::-1]: 
  print('{0:}({1:.3f})'.format(vocablist_DTM_tk2[word_num], coef))


부정적인 단어 Top 20 (낮은 평점과 상관관계가 강한 단어들)
최악(-3.353)
실망(-3.069)
비추(-2.724)
다시는(-2.653)
좋아요 좋아요(-2.469)
그래요(-2.245)
맛있어요 맛있어요(-2.202)
별로(-2.171)
약(-2.133)
심하네요(-2.094)
떨어져요(-2.084)
불만족(-2.073)
좋아요 아주(-2.047)
환불(-2.042)
맛없어요(-2.023)
안좋아요(-2.010)
안됩니다(-1.997)
불편(-1.993)
반품(-1.978)
안됨(-1.971)


## TFIDF + GridSearchCV + tokenizer_2
* TFIDF
* GridSearchCV
* Tokenizer2

In [237]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_vectorizer2 = TfidfVectorizer(ngram_range=(1,2), min_df = 3, max_df = 0.9, tokenizer = tokenizer)
tfidf_vectorizer2.fit(X_train_texts)
tfidf_matrix_train = tfidf_vectorizer2.transform(X_train_texts)
tfidf_matrix_test = tfidf_vectorizer2.transform(X_test_texts)

c:\Users\taeri\miniconda3\envs\ds_study\lib\site-packages\sklearn\feature_extraction\text.py:528: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(


In [238]:
vocablist_TFIDF_tk2 = [word for word, number in sorted(vect2.vocabulary_.items(), key = lambda x:x[1])]

In [239]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.model_selection import GridSearchCV

lr = LogisticRegression(random_state = 0, solver = 'liblinear', max_iter = 1000)
params = { 'C': [0.6, 0.8, 1 ,3.5, 4.5 ] }
grid_cv = GridSearchCV(lr , param_grid=params , cv=3 ,scoring='accuracy', verbose=1 )
grid_cv.fit(tfidf_matrix_train, y_train)
print(grid_cv.best_params_ , round(grid_cv.best_score_,4))

Fitting 3 folds for each of 5 candidates, totalling 15 fits
{'C': 3.5} 0.9125


In [240]:
best_estimator_TFIDF_tk2 = grid_cv.best_estimator_
y_pred = best_estimator_TFIDF_tk2.predict(X_test_tf)
print("accuracy: %.2f" % accuracy_score(y_test, y_pred))
print("Precision : %.3f" % precision_score(y_test, y_pred))
print("Recall : %.3f" % recall_score(y_test, y_pred))
print("F1 : %.3f" % f1_score(y_test, y_pred))

accuracy: 0.90
Precision : 0.884
Recall : 0.931
F1 : 0.907


In [241]:
import pickle

with open('TFIDF_tokenizer2_gridsearch_done_improved.pickle','wb') as fw:
    pickle.dump(best_estimator_TFIDF_tk2, fw)

In [242]:
coefficients = best_estimator_TFIDF_tk2.coef_.tolist()

sorted_coefficients = sorted(enumerate(coefficients[0]), key=lambda x:x[1], reverse=True)
# coefficients(계수)가 큰 값부터 내림차순으로 정렬

print('긍정적인 단어 Top 20 (높은 평점과 상관관계가 강한 단어들)')
for word_num, coef in sorted_coefficients[:20]:
  print('{0:}({1:.3f})'.format(vocablist_TFIDF_tk2[word_num], coef))

긍정적인 단어 Top 20 (높은 평점과 상관관계가 강한 단어들)
좋아요(14.662)
맛있어요(11.059)
만족해요(9.158)
예뻐요(8.888)
최고(8.511)
빠르고(7.974)
^^(7.707)
만족(7.237)
ㅎㅎ(7.068)
!(6.849)
~^^(6.684)
튼튼하고(6.457)
편(5.974)
괜찮네요(5.902)
너무 예뻐요(5.853)
항상(5.837)
확실히(5.750)
편하고(5.536)
:)(5.504)
없이 잘(5.492)


In [243]:
print('\n부정적인 단어 Top 10 (낮은 평점과 상관관계가 강한 단어들)')
for word_num, coef in sorted_coefficients[-20:][::-1]: 
  print('{0:}({1:.3f})'.format(vocablist_TFIDF_tk2[word_num], coef))


부정적인 단어 Top 10 (낮은 평점과 상관관계가 강한 단어들)
실망(-12.800)
별로(-11.660)
반품(-10.577)
최악(-10.169)
비추(-9.050)
그렇다(-8.953)
다시는(-8.436)
ㅡㅡ(-7.543)
불편(-7.196)
환불(-6.941)
좋아요 좋아요(-6.888)
빠르고 좋아요(-6.802)
안좋아요(-6.725)
그래요(-6.526)
불량(-6.373)
못(-6.363)
마세요(-6.294)
떨어져요(-6.152)
엉망(-6.012)
약(-5.981)


### DTM + 하이퍼파라미터 튜닝 + 유의어 일원화 보완(보완전) + 명사/형용사/동사만 추출
* accuracy: 0.90 (0.90)
* Precision : 0.905 (0.905)
* Recall : 0.891 (0.892)
* F1 : 0.898 (0.898)

### TFIDF + 하이퍼파라미터 튜닝 + 유의어 일원화 보완(보완전)+ 명사/형용사/동사만 추출
* accuracy: 0.90(0.90)
* Precision : 0.905(0.907)
* Recall : 0.893(0.894)
* F1 : 0.899(0.900)

### DTM + 하이퍼파라미터 튜닝 + 유의어 일원화 보완(보완전)+ 형태소만 추출
* accuracy: 0.91(0.91)
* Precision : 0.915(0.914)
* Recall : 0.910(0.910)
* F1 : 0.912 (0.912)

### TFIDF + 하이퍼파라미터 튜닝 + 유의어 일원화 보완(보완전)+ 형태소만 추출
* accuracy: 0.90(0.90)
* Precision : 0.884(0.883)
* Recall : 0.931(0.931)
* F1 : 0.907(0.906)